## Funktionen (Functions)

#### Warum brauchen wir Funktionen?

1. Um den geschriebenen Code in kleinere Unterteile zu zerlegen und die Lesbarkeit sowie Verständlichkeit des Codes zu erhöhen.
2. Um Codewiederholungen zu vermeiden, indem Abläufe, die sich im Programm mehrfach wiederholen, in einer einzigen Funktion zusammengefasst werden.

#### Definition einer Funktion

###### Syntax-Struktur:

`def funktions_name(funktions_argumente):`

`    operationen_der_funktion`
    
#### WICHTIG!
Nicht jede Funktion muss ein Argument entgegennehmen. Wenn sie keine Argumente akzeptiert, bleibt das Innere der Klammern leer. Ob eine Funktion Parameter benötigt oder nicht, hängt ganz von der Aufgabe ab, die sie erfüllen soll.

#### Parameterlose Funktion und Funktionsaufruf
Wenn eine Funktion keinen Parameter entgegennimmt, bleibt das Innere der Klammern sowohl bei der Definition als auch beim Aufruf leer.

In [2]:
def grussAusgeben():
    print("Hallo Welt!")
    print("Willkommen zum Python-Kurs")

grussAusgeben() # Funktionsaufruf

Hallo Welt!
Willkommen zum Python-Kurs


#### Funktionen mit Parametern
Wird verwendet, wenn die Funktion Operationen basierend auf Werten ausführen soll, die von außen übergeben werden.

In [5]:
def personalisierterGruss(name):
    print("Hallo {}, willkommen zum Kurs!".format(name))

personalisierterGruss("Max")
personalisierterGruss("Ahmet")

Hallo Max, willkommen zum Kurs!
Hallo Ahmet, willkommen zum Kurs!


In [6]:
def zweiZahlenAddieren(zahl1, zahl2):
    print("Die Summe der Zahlen lautet:", zahl1 + zahl2)

zweiZahlenAddieren(10, 15)

Die Summe der Zahlen lautet: 25


#### Das Konzept von `return` in Funktionen
Das Schlüsselwort `return` wird verwendet, um das Ergebnis der in der Funktion durchgeführten Operationen an die Stelle zurückzugeben, an der die Funktion aufgerufen wurde. Dadurch kann der resultierende Wert in einer Variablen gespeichert oder in anderen Berechnungen weiterverwendet werden.

In [5]:
def addierenMitReturn(zahl1, zahl2):
    summe = zahl1 + zahl2
    return summe

ergebnis = addierenMitReturn(10, 15)
print("Ergebnis aus der Funktion:", ergebnis)
print(ergebnis)
print(ergebnis**2)

Ergebnis aus der Funktion: 25
25
625


#### Standardparameter (Default-Werte) in Funktionen
Wenn für einen Parameter ein Standardwert definiert ist, ist die Übergabe dieses Arguments beim Funktionsaufruf optional. Wird kein Wert übergeben, verwendet die Funktion automatisch den vordefinierten Standardwert.

In [1]:
def benutzerInfoZeigen(name="Max", nachname="Mustermann", stadt="Berlin"):
    print("Benutzerinformationen -> Name:", name, "Nachname:", nachname, "Stadt:", stadt)

benutzerInfoZeigen()
benutzerInfoZeigen("Fatih", "Deniz")
benutzerInfoZeigen("Ahmet", "Yilmaz", "Istanbul")

Benutzerinformationen -> Name: Max Nachname: Mustermann Stadt: Berlin
Benutzerinformationen -> Name: Fatih Nachname: Deniz Stadt: Berlin
Benutzerinformationen -> Name: Ahmet Nachname: Yilmaz Stadt: Istanbul


#### Beliebige Anzahl von Argumenten (`*args`)
Wird verwendet, wenn vorab nicht bekannt ist, wie viele Argumente der Funktion übergeben werden. Die Argumente werden als ein Tupel (Tuple) zusammengefasst.

In [7]:
def beliebigeAddition(*zahlen):
    gesamtsumme = 0
    for zahl in zahlen:
        gesamtsumme += zahl
    return gesamtsumme

print(beliebigeAddition(1, 2, 3, 4, 5))
print(beliebigeAddition(10, 15, 20))
print(beliebigeAddition())

15
45
0


#### Rekursive Funktionen (Sich selbst aufrufende Funktionen)
Eine Funktion, die sich innerhalb ihres eigenen Codeblocks unter bestimmten Bedingungen selbst wieder aufruft, wird als rekursive Funktion bezeichnet.

In [3]:
def berechneFakultaet(zahl):
    if zahl == 1 or zahl == 0:
        return 1
    else:
        return zahl * berechneFakultaet(zahl - 1)

print(berechneFakultaet(5))

120


##### Funktionsweise der Fakultätsberechnung für `berechneFakultaet(5)`:

`berechneFakultaet(1)` Ergebnis -> `1`

`berechneFakultaet(2)` Ergebnis -> `2 * berechneFakultaet(1) = 2`

`berechneFakultaet(3)` Ergebnis -> `3 * berechneFakultaet(2) = 6`

`berechneFakultaet(4)` Ergebnis -> `4 * berechneFakultaet(3) = 24`

`berechneFakultaet(5)` Ergebnis -> `5 * berechneFakultaet(4) = 120`

#### Lokale und Globale Variablen
Variablen, die innerhalb einer Funktion definiert werden, sind **lokale Variablen** (local variables) und können nur innerhalb dieses Funktionsblocks verwendet werden.
Variablen, die im Hauptprogramm (Main) definiert werden, sind **globale Variablen** (global variables) und können in allen Codeblöcken des Programms aufgerufen werden.

In [4]:
def aendereZahl(x):
    print(y) # Zugriff auf die globale Variable y
    z = int() # Lokale Variable
    x = 5 # Lokale Variable

y = 50 # Globale Variable
aendereZahl(y)
print(y)

50
50
